### PR TITLE
Overviews for projects, dependencies, versions with filters

### PR DESCRIPTION
**Pull request by c0rydoras**
image
image
image
image
image
image
image
image

Once this is merged:
- make dependabot happy
- https://github.com/adfinis/Outdated/issues/149
- https://github.com/adfinis/Outdated/issues/153
- ...

### CODE REVIEW

File: ember/app/components/version-detailed.gjs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    <hr class='seperator' />
+    <div class='uk-margin-top'>
+      <ul
+        class='uk-list uk-flex uk-flex-between uk-flex-stretch uk-margin-remove'
+      >
+        {{#each this.cells as |cell|}}
+          <li class='uk-margin-remove'><strong
+              class='uk-display-block'
+            >{{cell.label}}</strong><span
```
Comment (derrabauke) [inline ember/app/components/version-detailed.gjs:91]: Could we *decamelize* those labels? `releaseVersion` doesn't look fancy.

File: ember/app/components/version-detailed.gjs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      </ul>
+    </div>
+    <hr class='uk-divider-icon' />
+    <Table @title='Used in' @data={{this.data}} as |t|>
+      <t.head />
+      <t.body />
+    </Table>
+    <EOLForm
+      @setIsEditing={{this.setIsEditing}}
+      @isEditing={{this.isEditing}}
```
Comment (derrabauke) [inline ember/app/components/version-detailed.gjs:104]: Those two actions are missing. :mag:

File: ember/app/controllers/dependencies/index.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  queryParams = ['name', 'provider'];
+
+  @tracked name = null;
+  @tracked provider;
+
+  @tracked model;
+
+  get providers() {
+    return ['PIP', 'NPM'];
+  }
```
Comment (derrabauke) [inline ember/app/controllers/dependencies/index.js:14]: Could become a future improvement: make the available providers configurable or request this information from the backend. Hardcoding this in some buried controller is not the way to go. :-)

File: ember/app/controllers/versions/index.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import _ from 'lodash';
```
Comment (derrabauke) [inline ember/app/controllers/versions/index.js:4]: https://www.npmjs.com/package/lodash.uniq

File: ember/app/models/dependency.js [hunk 5]
```diff
@@ -2,5 +2,14 @@ import Model, { attr } from '@ember-data/model';
 
 export default class DependencyModel extends Model {
   @attr name;
-  @attr status;
+  @attr provider;
+
+  get url() {
+    const providerMap = {
+      PIP: 'https://pypi.org/project/',
+      NPM: 'https://npmjs.com/package/',
```
Comment (derrabauke) [inline ember/app/models/dependency.js:10]: move to config if possible

File: ember/app/models/release-version.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
 import Model, { attr, belongsTo } from '@ember-data/model';
+import { tracked } from '@glimmer/tracking';
```
Comment (derrabauke) [inline ember/app/models/release-version.js:2]: As mentioned above: move intermediate state to modifying component.

File: ember/app/templates/dependencies/index.hbs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      @selected={{this.provider}}
+      @allowClear={{true}}
+      id="provider"
+      as |p|
+    >
+      {{p}}
+    </PowerSelect>
+
+  </label>
+</div>
```
Comment (derrabauke) [inline ember/app/templates/dependencies/index.hbs:27]: Could possibly become it's own component with some feature flags (status, provider, version, dependency). :man_shrugging:

File: ember/app/utils/status-to-class.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+export default function (status) {
+  const colorMappings = {
+    OUTDATED: 'danger',
+    WARNING: 'warning',
+    'UP-TO-DATE': 'success',
+  };
+  return colorMappings[status]
+    ? `uk-text-${colorMappings[status]}`
+    : 'text-undefined';
```
Comment (derrabauke) [inline ember/app/utils/status-to-class.js:9]: mentioned above: `.uk-text-muted`

**Review by derrabauke — COMMENTED** [review]
This PR is pretty huge and was a pain in the :peach: to review. Please try to make smaller contributions if possible next time. ;-)

File: ember/.stylelintrc.js [hunk 9]
```diff
@@ -5,5 +5,6 @@ module.exports = {
   extends: ['stylelint-prettier/recommended', 'stylelint-config-standard-scss'],
   rules: {
     'scss/at-extend-no-missing-placeholder': null,
+    'selector-class-pattern': null,
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/.stylelintrc.js:8]:
```
[lint:scss]  28:3  ✖  Expected class selector ".uk-flex-row\@s" to be kebab-case  selector-class-pattern
[lint:scss]  32:3  ✖  Expected class selector ".gap\@s" to be kebab-case          selector-class-pattern
```

It throws errors when using `\@{size}`

File: ember/app/components/version-detailed.gjs (hunk 1 above)
Comment (c0rydoras) [reply ember/app/components/version-detailed.gjs:91]: those are decamelized by this `startCase`
https://github.com/adfinis/Outdated/blob/7962d572dcb4209a9ebc65de6cfbb6ada7e708d7/ember/app/components/version-detailed.gjs#L62

image

File: ember/app/routes/projects/index.js [hunk 10]
```diff
@@ -4,11 +4,11 @@ import { service } from '@ember/service';
 export default class ProjectsRoute extends Route {
   @service store;
 
-  beforeModel() {
-    const unsavedProject = this.store
+  async beforeModel() {
+    const unsavedProject = await this.store
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/routes/projects/index.js:8]: AFAIK destroyRecord is not

https://api.emberjs.com/ember-data/5.3/classes/Model/methods/destroyRecord?anchor=destroyRecord

File: ember/package.json [hunk 11]
```diff
@@ -71,14 +74,14 @@
     "ember-simple-auth": "^6.0.0",
     "ember-simple-auth-oidc": "^6.0.0",
     "ember-source": "^5.4.0",
-    "ember-template-lint": "^5.11.2",
-    "ember-template-lint-plugin-prettier": "^5.0.0",
+    "ember-template-imports": "^4.0.0",
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/package.json:77]: i don't see anything that was downgraded

File: ember/app/styles/components/_text.scss [hunk 12]
```diff
@@ -1,19 +1,16 @@
 .text-undefined {
-  &:not(tr),
-  td[data-test-version-end-of-life] {
-    @extend .uk-text-muted;
+  @extend .uk-text-muted;
+}
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/styles/components/_text.scss]: fixed here: https://github.com/adfinis/Outdated/pull/170/commits/23ace0a40958614a642fc0a95ea57da03046c587

File: ember/app/helpers/status-to-class.js [hunk 13]
```diff
@@ -1,12 +1,5 @@
...
 
-export default helper(function statusToClass(status) {
-  const colorMappings = {
-    OUTDATED: 'danger',
-    WARNING: 'warning',
-    'UP-TO-DATE': 'success',
-  };
-  return colorMappings[status]
-    ? `uk-text-${colorMappings[status]}`
-    : 'text-undefined';
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/helpers/status-to-class.js:11]: fixed here: https://github.com/adfinis/Outdated/pull/170/commits/23ace0a40958614a642fc0a95ea57da03046c587

File: ember/app/utils/status-to-class.js (hunk 8 above)
Comment (c0rydoras) [reply ember/app/utils/status-to-class.js]: fixed here: https://github.com/adfinis/Outdated/pull/170/commits/23ace0a40958614a642fc0a95ea57da03046c587

File: ember/app/models/release-version.js (hunk 6 above)
Comment (c0rydoras) [reply ember/app/models/release-version.js:2]: im not sure how to then properly implement that with my current form validation

File: ember/app/components/eol-form.gjs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  saveReleaseVersion = dropTask(async (data) => {
+    try {
+      const changes = await data.changes;
+      if (changes.length) {
+        const releaseVersion = this.args.version.releaseVersion;
+        releaseVersion.endOfLife = changes[0].value;
+        await releaseVersion.save();
+      }
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/components/eol-form.gjs:29]: im not sure how to then properly implement that with my current form validation

File: ember/app/components/table.gjs [hunk 15]
```diff
@@ -0,0 +1,103 @@
+import { hash, concat } from '@ember/helper';
+import { service } from '@ember/service';
+import { htmlSafe } from '@ember/template';
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import formatDate from 'ember-intl/helpers/format-date';
+import { or } from 'ember-truth-helpers';
+import _ from 'lodash';
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/components/table.gjs]: done here: https://github.com/adfinis/Outdated/pull/170/commits/b6ff8b6ba2170c787246aac1e1ece093ba0778d6

File: ember/app/controllers/versions/index.js (hunk 4 above)
Comment (c0rydoras) [reply ember/app/controllers/versions/index.js]: done here: https://github.com/adfinis/Outdated/pull/170/commits/b6ff8b6ba2170c787246aac1e1ece093ba0778d6

File: ember/app/components/table.gjs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      ...new Set(
+        this.args.fields ??
+          this.args.data.map((object) => Object.keys(object.values)).flat(),
+      ),
+    ];
+  }
+
+  get hasData() {
+    return !!this.args.data?.length;
+  }
```
[the remark this answers is not in the document]
Comment (c0rydoras) [reply ember/app/components/table.gjs]: done here: resolved

File: ember/app/components/version-detailed.gjs (hunk 2 above)
Comment (c0rydoras) [reply ember/app/components/version-detailed.gjs]: fixed here: https://github.com/adfinis/Outdated/pull/170/commits/3d2d83400995ad819b4e219acfdd951219b0aa1a

File: ember/app/controllers/dependencies/index.js (hunk 3 above)
Comment (c0rydoras) [reply ember/app/controllers/dependencies/index.js:14]: I created an issue for it: https://github.com/adfinis/Outdated/issues/186

File: ember/app/models/dependency.js (hunk 5 above)
Comment (c0rydoras) [reply ember/app/models/dependency.js:10]: as specified in [here](https://github.com/adfinis/Outdated/issues/186) this should also be returned by the backend

File: ember/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (derrabauke) [reply ember/package.json:77]: Getting old. Sorry bad eyes. :older_man:

File: ember/app/routes/projects/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (derrabauke) [reply ember/app/routes/projects/index.js:8]: for sure `destroyRecord` is async. But the call before doesn't need the `await`. More or less a nitpick.
```suggestion
    const unsavedProject = this.store
```

File: ember/app/components/version-detailed.gjs (hunk 1 above)
Comment (derrabauke) [reply ember/app/components/version-detailed.gjs:90]: sweet :ok_hand:

File: ember/app/templates/dependencies/index.hbs (hunk 7 above)
Comment (c0rydoras) [reply ember/app/templates/dependencies/index.hbs]: Done here: https://github.com/adfinis/Outdated/pull/170/commits/8ed9cef9776f2be5597fdea40db24c486e5b68b7

File: ember/app/models/release-version.js (hunk 6 above)
Comment (derrabauke) [reply ember/app/models/release-version.js:2]: Well, then leave it as is for right now.

File: ember/app/templates/dependencies/index.hbs (hunk 7 above)
Comment (derrabauke) [reply ember/app/templates/dependencies/index.hbs]: Nice job! :+1:

File: ember/.prettierrc.js [hunk 17]
```diff
@@ -1,9 +1,10 @@
 'use strict';
 
 module.exports = {
+  plugins: ['prettier-plugin-ember-template-tag'],
   overrides: [
     {
-      files: '*.{js,ts}',
+      files: '*.{js,ts,gjs,gts}',
```
[the remark this answers is not in the document]
Comment (derrabauke) [reply ember/.prettierrc.js:7]: I guess so, since it's your apprenticeship project. :shrug:

**Review by derrabauke — APPROVED** [review]

**Pull request merged**